qiskit_nature/second_q/operators/polynomial_tensor.py

```diff
@@ -0,0 +1,126 @@
+        copy_dict: Dict[str, np.ndarray] = {}
+
+        shapes = set()
+        for key, value in data.items():
+            if isinstance(value, Number):
+                value = np.asarray(value)
+
+            if len(value.shape) != len(key):
+                raise ValueError(
+                    f"Data key {key} of length {len(key)} does not match "
+                    f"data value matrix of dimensions {value.shape}"
+                )
+
+            dims = set(value.shape)
+
+            if len(dims) > 1:
+                raise ValueError(
+                    f"For key {key}: dimensions of value matrix are not identical {value.shape}"
+                )
+
+            shapes.update(dims)
+            copy_dict[key] = value
+
+        if len(shapes) != 1:
+            raise ValueError("Dimensions of value matrices in data dictionary are not identical.")
+
+        self._data = copy_dict
+
+    def mul(self, other: complex):
+        """Scalar multiplication of PolynomialTensor with complex"""
+
+        if not isinstance(other, Number):
+            raise TypeError(f"other {other} must be a number")
+
+        prod_dict: Dict[str, np.ndarray] = {}
+        for key, matrix in self._data.items():
+            prod_dict[key] = np.multiply(matrix, other)
+        return PolynomialTensor(prod_dict)
+
+    def _multiply(self, other):
+        return self.mul(other)
+
+    def add(self, other: PolynomialTensor):
+        """Addition of PolynomialTensors"""
+
+        if not isinstance(other, PolynomialTensor):
+            raise TypeError("Incorrect argument type: other should be PolynomialTensor")
+
+        sum_dict: Dict[str, np.ndarray] = self._data.copy()
+        for other_key, other_value in other._data.items():
+            if other_key in sum_dict.keys():
+                if other_value.shape == np.shape(sum_dict[other_key]):
+                    sum_dict[other_key] = np.add(other_value, sum_dict[other_key])
+                else:
+                    raise ValueError(
+                        f"For key {other_key}: "
+                        f"corresponding data value of shape {np.shape(sum_dict[other_key])} "
+                        f"does not match other value matrix of shape {other_value.shape}"
+                    )
+            else:
+                sum_dict[other_key] = other_value
+
+        return PolynomialTensor(sum_dict)
+
+    def _add(self, other, qargs=None):
+        return self.add(other)
+
+    def __eq__(self, other):
+        """Check equality of PolynomialTensors"""
+
+        if not isinstance(other, PolynomialTensor):
+            raise TypeError("Incorrect argument type: other should be PolynomialTensor")
+
+        if self._data.keys() != other._data.keys():
+            return False
+        for key, value in self._data.items():
+            if not np.allclose(value, other._data[key], atol=self.atol, rtol=self.rtol):
+                return False
```

Comment:
This is actually the implementation of the `equiv` method.

The `__eq__` method should check for exact equality. It would be good to implement both